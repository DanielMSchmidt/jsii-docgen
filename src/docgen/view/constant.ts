import * as reflect from 'jsii-reflect';
import { Markdown } from '../render/markdown';
import { PropertySchema } from '../schema';
import { Transpile, TranspiledType } from '../transpile/transpile';
import { Property } from './property';

export class Constant {
  private readonly constant: Property;
  constructor(transpile: Transpile, property: reflect.Property, linkFormatter: (type: TranspiledType) => string) {
    this.constant = new Property(transpile, property, linkFormatter);
  }
  public toMarkdown(): Markdown {
    return this.constant.toMarkdown();
  }

  public toJson(): PropertySchema {
    return this.constant.toJson();
  }
}
