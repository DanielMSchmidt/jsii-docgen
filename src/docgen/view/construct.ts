import * as reflect from 'jsii-reflect';
import { Markdown } from '../render/markdown';
import { ConstructSchema } from '../schema';
import { Transpile, TranspiledType } from '../transpile/transpile';
import { Class } from './class';

export class Construct {
  private readonly construct: Class;
  constructor(transpile: Transpile, klass: reflect.ClassType, linkFormatter: (type: TranspiledType) => string) {
    this.construct = new Class(transpile, klass, linkFormatter);
  }
  public toMarkdown(): Markdown {
    return this.construct.toMarkdown();
  }

  public toJson(): ConstructSchema {
    return this.construct.toJson();
  }
}
