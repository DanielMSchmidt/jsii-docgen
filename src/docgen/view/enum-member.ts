import * as reflect from 'jsii-reflect';
import { EnumMemberSchema, extractDocs } from '../schema';
import { Transpile, TranspiledEnumMember } from '../transpile/transpile';

export class EnumMember {
  private readonly transpiled: TranspiledEnumMember;
  constructor(transpile: Transpile, private readonly em: reflect.EnumMember) {
    this.transpiled = transpile.enumMember(em);
  }

  public toJson(): EnumMemberSchema {
    return {
      jsiiId: `${this.em.enumType.fqn}.${this.em.name}`,
      displayName: this.transpiled.name,
      fqn: this.transpiled.fqn,
      docs: extractDocs(this.em.docs),
      submoduleJsiiId: this.transpiled.type.submoduleJsiiId,
      submoduleFqn: this.transpiled.type.submoduleFqn,
      sourceFile: this.transpiled.type.source.locationInModule?.filename,
      sourceLine: this.transpiled.type.source.locationInModule?.line,
    };
  }
}
