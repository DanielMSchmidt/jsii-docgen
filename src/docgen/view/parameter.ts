import * as reflect from 'jsii-reflect';
import { Markdown } from '../render/markdown';
import { ParameterSchema } from '../schema';
import { Transpile, TranspiledParameter, TranspiledType } from '../transpile/transpile';

export class Parameter {
  private readonly transpiled: TranspiledParameter;
  constructor(
    transpile: Transpile,
    private readonly parameter: reflect.Parameter,
  ) {
    this.transpiled = transpile.parameter(parameter);
  }

  public toMarkdown(linkFormatter: (type: TranspiledType) => string): Markdown {
    const optionality = this.parameter.optional ? 'Optional' : 'Required';

    const md = new Markdown({
      id: `${this.transpiled.parentType.fqn}.parameter.${this.transpiled.name}`,
      header: {
        title: this.transpiled.name,
        sup: optionality,
        pre: true,
        strike: this.parameter.docs.deprecated,
      },
    });

    if (this.parameter.docs.deprecated) {
      md.bullet(
        `${Markdown.italic('Deprecated:')} ${
          this.parameter.docs.deprecationReason
        }`,
      );
      md.lines('');
    }

    const metadata: any = {
      Type: this.transpiled.typeReference.toString({
        typeFormatter: (t) => `[${Markdown.pre(t.fqn)}](${linkFormatter(t)})`,
        stringFormatter: Markdown.pre,
      }),
    };

    if (this.parameter.spec.docs?.default) {
      metadata.Default = Markdown.sanitize(this.parameter.spec.docs?.default);
    }

    for (const [key, value] of Object.entries(metadata)) {
      md.bullet(`${Markdown.italic(`${key}:`)} ${value}`);
    }
    md.lines('');

    if (this.parameter.docs) {
      md.docs(this.parameter.docs);
    }

    md.split();

    return md;
  }

  public toJson(): ParameterSchema {
    return {
      id: `${this.transpiled.parentType.fqn}.${this.transpiled.name}`,
      name: this.transpiled.name,
      optional: this.parameter.optional,
      deprecated: this.parameter.docs.deprecated,
      deprecationReason: this.parameter.docs.deprecationReason,
      docs: this.parameter.docs.toString(),
      default: this.parameter.spec.docs?.default,
      type: this.transpiled.typeReference.toJson(),
    };
  }
}
