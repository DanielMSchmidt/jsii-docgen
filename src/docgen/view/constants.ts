import * as reflect from 'jsii-reflect';
import { Markdown } from '../render/markdown';
import { PropertySchema } from '../schema';
import { Transpile, TranspiledType } from '../transpile/transpile';
import { Constant } from './constant';

export class Constants {
  private readonly constants: Constant[];
  constructor(transpile: Transpile, properties: reflect.Property[]) {
    this.constants = properties
      .filter((p) => !p.protected && p.const)
      .map((p) => new Constant(transpile, p));
  }

  public toMarkdown(linkFormatter: (type: TranspiledType) => string): Markdown {
    if (this.constants.length === 0) {
      return Markdown.EMPTY;
    }

    const md = new Markdown({ header: { title: 'Constants' } });
    for (const c of this.constants) {
      md.section(c.toMarkdown(linkFormatter));
    }
    return md;
  }

  public toJson(): PropertySchema[] {
    return this.constants.map((constant) => constant.toJson());
  }
}
