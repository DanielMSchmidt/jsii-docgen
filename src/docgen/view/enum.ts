import * as reflect from 'jsii-reflect';
import { Markdown } from '../render/markdown';
import { EnumSchema } from '../schema';
import { Transpile } from '../transpile/transpile';
import { EnumMember } from './enum-member';

export class Enum {
  private readonly members: EnumMember[];
  constructor(
    private readonly transpile: Transpile,
    private readonly enu: reflect.EnumType,
  ) {
    this.members = enu.members.map((em) => new EnumMember(transpile, em));
  }

  public toMarkdown(): Markdown {
    const transpiled = this.transpile.enum(this.enu);
    const md = new Markdown({ header: { title: transpiled.name } });

    if (this.enu.docs) {
      md.docs(this.enu.docs);
    }

    for (const m of this.members) {
      md.section(m.toMarkdown());
    }

    return md;
  }

  public toJson(): EnumSchema {
    const transpiled = this.transpile.enum(this.enu);
    return {
      name: transpiled.name,
      docs: this.enu.docs.toString(),
      members: this.members.map((member) => member.toJson()),
    };
  }
}
