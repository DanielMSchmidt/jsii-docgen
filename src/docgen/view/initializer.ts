import * as reflect from 'jsii-reflect';
import { Markdown } from '../render/markdown';
import { InitializerSchema } from '../schema';
import { Transpile, TranspiledCallable, TranspiledType } from '../transpile/transpile';
import { Parameter } from './parameter';

export class Initializer {
  private readonly transpiled: TranspiledCallable;
  private readonly parameters: Parameter[];
  constructor(
    private readonly transpile: Transpile,
    initializer: reflect.Initializer,
  ) {
    this.transpiled = transpile.callable(initializer);
    this.parameters = this.transpiled.parameters.map(
      (p) => new Parameter(this.transpile, p),
    );
  }

  public toMarkdown(
    linkFormatter: (type: TranspiledType) => string,
  ): Markdown {
    const md = new Markdown({
      id: `${this.transpiled.parentType.fqn}.Initializer`,
      header: {
        title: 'Initializers',
      },
    });

    md.code(
      this.transpile.language.toString(),
      `${this.transpiled.import}`,
      '',
      ...this.transpiled.invocations,
    );

    for (const parameter of this.parameters) {
      md.section(parameter.toMarkdown(linkFormatter));
    }

    return md;
  }

  public toJson(): InitializerSchema {
    const md = new Markdown();
    md.code(
      this.transpile.language.toString(),
      `${this.transpiled.import}`,
      '',
      ...this.transpiled.invocations,
    );

    return {
      id: `${this.transpiled.parentType.fqn}.Initializer`,
      snippet: md.render(),
      parameters: this.parameters.map((parameter) => parameter.toJson()),
    };
  }
}
