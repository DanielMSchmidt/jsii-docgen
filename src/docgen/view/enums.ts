import * as reflect from 'jsii-reflect';
import { Markdown } from '../render/markdown';
import { EnumSchema } from '../schema';
import { Transpile } from '../transpile/transpile';
import { Enum } from './enum';

export class Enums {
  private readonly enums: Enum[];
  constructor(transpile: Transpile, enums: reflect.EnumType[]) {
    this.enums = enums.map((e) => new Enum(transpile, e));
  }

  public toMarkdown(): Markdown {
    if (this.enums.length === 0) {
      return Markdown.EMPTY;
    }

    const md = new Markdown({ header: { title: 'Enums' } });
    for (const e of this.enums) {
      md.section(e.toMarkdown());
    }
    return md;
  }

  public toJson(): EnumSchema[] {
    return this.enums.map((e) => e.toJson());
  }
}
