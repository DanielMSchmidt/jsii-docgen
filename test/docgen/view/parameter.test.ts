import * as reflect from 'jsii-reflect';
import { CSharpTranspile } from '../../../src/docgen/transpile/csharp';
import { JavaTranspile } from '../../../src/docgen/transpile/java';
import { PythonTranspile } from '../../../src/docgen/transpile/python';
import { TranspiledType } from '../../../src/docgen/transpile/transpile';
import { TypeScriptTranspile } from '../../../src/docgen/transpile/typescript';
import { Parameter } from '../../../src/docgen/view/parameter';
import { Assemblies } from '../assemblies';

const assembly: reflect.Assembly = Assemblies.instance.withoutSubmodules;

const findParameter = (): reflect.Parameter => {
  for (const klass of assembly.system.classes) {
    for (const method of klass.ownMethods) {
      if (method.parameters.length > 0) {
        return method.parameters[0];
      }
    }
  }
  throw new Error('Assembly does not contain a parameter');
};

describe('python', () => {
  const transpile = new PythonTranspile();
  test('markdown snapshot', () => {
    const parameter = new Parameter(transpile, findParameter());
    expect(parameter.toMarkdown((t: TranspiledType) => `#${t.fqn}`).render()).toMatchSnapshot();
  });

  test('json snapshot', () => {
    const parameter = new Parameter(transpile, findParameter());
    expect(parameter.toJson()).toMatchSnapshot();
  });
});

describe('typescript', () => {
  const transpile = new TypeScriptTranspile();
  test('markdown snapshot', () => {
    const parameter = new Parameter(transpile, findParameter());
    expect(parameter.toMarkdown((t: TranspiledType) => `#${t.fqn}`).render()).toMatchSnapshot();
  });

  test('json snapshot', () => {
    const parameter = new Parameter(transpile, findParameter());
    expect(parameter.toJson()).toMatchSnapshot();
  });
});

describe('java', () => {
  const transpile = new JavaTranspile();
  test('markdown snapshot', () => {
    const parameter = new Parameter(transpile, findParameter());
    expect(parameter.toMarkdown((t: TranspiledType) => `#${t.fqn}`).render()).toMatchSnapshot();
  });

  test('json snapshot', () => {
    const parameter = new Parameter(transpile, findParameter());
    expect(parameter.toJson()).toMatchSnapshot();
  });
});

describe('csharp', () => {
  const transpile = new CSharpTranspile();
  test('markdown snapshot', () => {
    const parameter = new Parameter(transpile, findParameter());
    expect(parameter.toMarkdown((t: TranspiledType) => `#${t.fqn}`).render()).toMatchSnapshot();
  });

  test('json snapshot', () => {
    const parameter = new Parameter(transpile, findParameter());
    expect(parameter.toJson()).toMatchSnapshot();
  });
});
