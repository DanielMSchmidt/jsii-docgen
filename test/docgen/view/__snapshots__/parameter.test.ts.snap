// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`csharp snapshot 1`] = `
Object {
  "default": undefined,
  "displayName": "Grantee",
  "docs": Object {},
  "fqn": "Amazon.CDK.AWS.ECR.AuthorizationToken.GrantRead.parameter.Grantee",
  "jsiiId": "@aws-cdk/aws-ecr.AuthorizationToken.grantRead.parameter.grantee",
  "optional": undefined,
  "sourceFile": "lib/auth-token.ts",
  "sourceLine": 8,
  "submoduleFqn": undefined,
  "submoduleJsiiId": undefined,
  "type": Object {
    "formattingPattern": "%",
    "types": Array [
      Object {
        "displayName": "IGrantable",
        "fqn": "Amazon.CDK.AWS.IAM.IGrantable",
        "jsiiId": "@aws-cdk/aws-iam.IGrantable",
        "jsiiProjectName": "@aws-cdk/aws-iam",
        "jsiiProjectVersion": "99.99.99",
        "packageName": "@aws-cdk/aws-iam",
        "packageVersion": "99.99.99",
      },
    ],
  },
}
`;

exports[`csharp snapshot 2`] = `
" \`Grantee\`<sup>Required</sup> <a name=\\"Grantee\\" id=\\"@aws-cdk/aws-ecr.AuthorizationToken.grantRead.parameter.grantee\\"></a>

- *Type:* undefined

---
"
`;

exports[`go snapshot 1`] = `
Object {
  "default": undefined,
  "displayName": "grantee",
  "docs": Object {},
  "fqn": "github.com/aws/aws-cdk-go/awscdkawsecr.AuthorizationToken.GrantRead.parameter.grantee",
  "jsiiId": "@aws-cdk/aws-ecr.AuthorizationToken.grantRead.parameter.grantee",
  "optional": undefined,
  "sourceFile": "lib/auth-token.ts",
  "sourceLine": 8,
  "submoduleFqn": undefined,
  "submoduleJsiiId": undefined,
  "type": Object {
    "formattingPattern": "%",
    "types": Array [
      Object {
        "displayName": "IGrantable",
        "fqn": "github.com/aws/aws-cdk-go/awscdk/awsiam/awscdkawsiam.IGrantable",
        "jsiiId": "@aws-cdk/aws-iam.IGrantable",
        "jsiiProjectName": "@aws-cdk/aws-iam",
        "jsiiProjectVersion": "99.99.99",
        "packageName": "@aws-cdk/aws-iam",
        "packageVersion": "99.99.99",
      },
    ],
  },
}
`;

exports[`go snapshot 2`] = `
" \`grantee\`<sup>Required</sup> <a name=\\"grantee\\" id=\\"@aws-cdk/aws-ecr.AuthorizationToken.grantRead.parameter.grantee\\"></a>

- *Type:* undefined

---
"
`;

exports[`java snapshot 1`] = `
Object {
  "default": undefined,
  "displayName": "grantee",
  "docs": Object {},
  "fqn": "software.amazon.awscdk.services.ecr.AuthorizationToken.grantRead.parameter.grantee",
  "jsiiId": "@aws-cdk/aws-ecr.AuthorizationToken.grantRead.parameter.grantee",
  "optional": undefined,
  "sourceFile": "lib/auth-token.ts",
  "sourceLine": 8,
  "submoduleFqn": undefined,
  "submoduleJsiiId": undefined,
  "type": Object {
    "formattingPattern": "%",
    "types": Array [
      Object {
        "displayName": "IGrantable",
        "fqn": "software.amazon.awscdk.services.iam.IGrantable",
        "jsiiId": "@aws-cdk/aws-iam.IGrantable",
        "jsiiProjectName": "@aws-cdk/aws-iam",
        "jsiiProjectVersion": "99.99.99",
        "packageName": "@aws-cdk/aws-iam",
        "packageVersion": "99.99.99",
      },
    ],
  },
}
`;

exports[`java snapshot 2`] = `
" \`grantee\`<sup>Required</sup> <a name=\\"grantee\\" id=\\"@aws-cdk/aws-ecr.AuthorizationToken.grantRead.parameter.grantee\\"></a>

- *Type:* undefined

---
"
`;

exports[`newlines in "defaults" are removed 1`] = `
" \`grantee\`<sup>Required</sup> <a name=\\"grantee\\" id=\\"@aws-cdk/aws-ecr.AuthorizationToken.grantRead.parameter.grantee\\"></a>

- *Type:* undefined
- *Default:* default option with a newline

---
"
`;

exports[`python snapshot 1`] = `
Object {
  "default": undefined,
  "displayName": "grantee",
  "docs": Object {},
  "fqn": "aws_cdk.aws_ecr.AuthorizationToken.grant_read.parameter.grantee",
  "jsiiId": "@aws-cdk/aws-ecr.AuthorizationToken.grantRead.parameter.grantee",
  "optional": undefined,
  "sourceFile": "lib/auth-token.ts",
  "sourceLine": 8,
  "submoduleFqn": undefined,
  "submoduleJsiiId": undefined,
  "type": Object {
    "formattingPattern": "%",
    "types": Array [
      Object {
        "displayName": "IGrantable",
        "fqn": "aws_cdk.aws_iam.IGrantable",
        "jsiiId": "@aws-cdk/aws-iam.IGrantable",
        "jsiiProjectName": "@aws-cdk/aws-iam",
        "jsiiProjectVersion": "99.99.99",
        "packageName": "@aws-cdk/aws-iam",
        "packageVersion": "99.99.99",
      },
    ],
  },
}
`;

exports[`python snapshot 2`] = `
" \`grantee\`<sup>Required</sup> <a name=\\"grantee\\" id=\\"@aws-cdk/aws-ecr.AuthorizationToken.grantRead.parameter.grantee\\"></a>

- *Type:* undefined

---
"
`;

exports[`typescript snapshot 1`] = `
Object {
  "default": undefined,
  "displayName": "grantee",
  "docs": Object {},
  "fqn": "@aws-cdk/aws-ecr.AuthorizationToken.grantRead.parameter.grantee",
  "jsiiId": "@aws-cdk/aws-ecr.AuthorizationToken.grantRead.parameter.grantee",
  "optional": undefined,
  "sourceFile": "lib/auth-token.ts",
  "sourceLine": 8,
  "submoduleFqn": undefined,
  "submoduleJsiiId": undefined,
  "type": Object {
    "formattingPattern": "%",
    "types": Array [
      Object {
        "displayName": "IGrantable",
        "fqn": "@aws-cdk/aws-iam.IGrantable",
        "jsiiId": "@aws-cdk/aws-iam.IGrantable",
        "jsiiProjectName": "@aws-cdk/aws-iam",
        "jsiiProjectVersion": "99.99.99",
        "packageName": "@aws-cdk/aws-iam",
        "packageVersion": "99.99.99",
      },
    ],
  },
}
`;

exports[`typescript snapshot 2`] = `
" \`grantee\`<sup>Required</sup> <a name=\\"grantee\\" id=\\"@aws-cdk/aws-ecr.AuthorizationToken.grantRead.parameter.grantee\\"></a>

- *Type:* undefined

---
"
`;
